### PR TITLE
Cleanup lower_generics_mut and make span be the bound itself

### PR DESCRIPTION
**Pull request by jackh726**
Closes #86298 (supersedes those changes)

r? @cjgillot since you reviewed the other PR

(Used wrong branch for #89338)

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/item.rs [hunk 1]
```diff
@@ -1356,32 +1356,46 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
+            };
+            let mut is_param: Option<bool> = None;
+            for bound in &bound_pred.bounds {
+                if !matches!(*bound, GenericBound::Trait(_, TraitBoundModifier::Maybe)) {
+                    continue;
+                }
+                // We only need to compute this once per `WherePredicate`, but don't
+                // need to compute this at all unless there is a Maybe bound.
+                // This closure should be able to be moved out of the loop,
+                // but `get_partial_res` takes a `&mut self`
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/item.rs:1371]: `get_partial_res` should take `&self`, it does not need to be mutable.

Comment (jackh726) [reply compiler/rustc_ast_lowering/src/item.rs:1371]: Yeah, I wasn't sure if it was worth changing that for; I wasn't sure if there's some design reason that it's `&mut self`. I'll go ahead and do that though.

**Comment by jackh726** [comment]
@cjgillot updated

**Pull request merged**